lib/charms/loki_k8s/v0/loki_push_api.py

```diff
@@ -1005,12 +1063,11 @@ def _labeled_alert_groups(self) -> list:
             message = "Failed to read alert rules (must contain {}): ".format(
                 ", ".join(must_contain)
             ) + ", ".join(map(str, invalid_files))
-            self._charm.model.unit.status = BlockedStatus(message)
+            raise AlertRulesError(message)
```

Comment:
Note that this is a potentially significant change of contract:
Before, the function would have returned the valid alert groups even if some are invalid; now nothing would be returned.

Reply:
cc @mmanciop

Reply:
We need to keep the alert groups but append to them the juju topology to avoid clashes with multiple deployments of the same charm

Reply:
@simskij @sed-i I refactored some methods, keeping the original contract. Now custom events are emitted from the same place (`_on_logging_relation_changed`) so it's easy to follow.